[docs] Getting started review (#165)
cozystack/website: content/en/docs/getting-started/hardware-requirements.md

weight: 5
---

Cozystack utilizes Talos Linux, a minimalistic Linux distribution designed solely to run Kubernetes.
Usually it means that it will **not** be possible to share a server with services other than Cozystack runs.
The good news is that whichever service you need, Cozystack will run it prefectly: securely, efficiently, and
in a fully containerized or virtualized environment.

Hardware requirements depend on your usage scenario.
Below are several common deployment options, review them to determine which setup fits best to your needs.

## Small lab

Here are the baseline requirements for running a tiny installation.

**Compute:**

- Three physical or virtual servers with amd64/x86_64 architecture, with at least 16 GB RAM and 4 CPU cores each.
- Virtualized servers need nested virtualization enabled and the CPU model set to `host` (without emulation).

> - PXE installation requires an extra management instance connected to the same network, with any Linux system able to run a Docker container.

Reviewer: ```suggestion
- PXE installation requires an extra management VM or physical server connected to the same network,
  with any Linux system able to run a Docker container.
```

Reviewer: We refer to VM in the following list item. Let's keep it explicit that VM and physical server are valid options.

Reviewer: After re-reading it for the fifth time, I finally understand what you mean. It shouldn't be a fresh VM or physical server. Most likely, it will be the regular workstation of an engineer who's doing the deployment.

  It should also have `x86-64-v2` architecture, which most probably may be achieved by setting CPU model to `host` in case of a VM.

**Storage:**

> - Primary disk should be at least 32 GB. Talos Linux uses it for the OS, images, and logs. It would be nice to have more.
> - Secondary disk is dedicated to the storage system. 100 GB is a good starting point, but the actual size depends entirely on your workload.

Reviewer: ```suggestion
The primary disk should be at least 32 GB. Talos Linux uses it for the OS, images, and logs, so it's better to have more space.
- Secondary disk is dedicated to the storage system. 100 GB is a good starting point, but the actual size depends entirely on your workload.
```

Reviewer: "It would be nice" is too indirect. Let's be more blunt when talking about hardware requirements.

**Networking:**

> - Machines should be able to use additional IPs, or an external load balancer should be available.
>   In public clouds the use of the additional IPs typically needs to be allowed explicitly.
> - Additional public IPs for ingress and virtual machines may be needed. Check if your public cloud provider supports floating IPs.

Reviewer: ```suggestion
- Machines must be allowed to use additional IPs, or an external load balancer must be available.
  Using additional IPs is disabled by default and must be enabled explicitly in most public clouds.
- Additional public IPs for ingress and virtual machines may be needed. Check if your public cloud provider supports floating IPs.
```

Reviewer: There's this important context about the option being available but disabled by default, which I don't want to lose.

## Production Cluster

For a production environment, consider the following:

**Compute:**

> - **Three worker nodes** are a mandatory for running highly-available applications.
>   If one of the three nodes becomes unavailable, either due to hardware failure or maintenance, you’ll be operating in a degraded state.
>   While database clusters and replicated storage will continue functioning, it won’t be possible to start new database instances or create replicated volumes.
> - Having separate servers for Kubernetes master nodes is highly recommended, although not required.

Reviewer: ```suggestion
- Having at least **three worker nodes** is mandatory for running highly available applications.
  If one of the three nodes becomes unavailable due to hardware failure or maintenance, you’ll be operating in a degraded state.
  While database clusters and replicated storage will continue functioning, starting new database instances or creating replicated volumes won’t be possible.
- Having separate servers for Kubernetes master nodes is highly recommended, although not required.
```

**Networking:**

- Distributed cluster requires both fast and reliable network, and it **must** have low RTT (Round Trip Time), as
  Kubernetes is not designed to operate efficiently over high-latency networks.

  Datacenters in the same city typically have less than 1 ms latency, which is ideal.
  The *maximum acceptable* RTT is 10 ms.
  Running Kubernetes or replicated storage over a network with RTT above 20 ms is strongly discouraged.
  To measure actual RTT you can use the `ping` command.

- It's also recommended to have at least 2–3 nodes per datacenter in a distributed cluster.
  This ensures that the cluster would be able to survive one datacenter loss without major disruption.

- If it's hard to keep a single address space between datacenters, instead of using some external VPN,
  you can enable **KubeSpan**, a Talos Linux feature that creates a WireGuard-backed full-mesh VPN between nodes.

> ## Highly-Available applications notes

Reviewer: I agree, it looks like just notes or additions to the previous chapters.

Should we merge it to Production? It has a note on HA in the Compute section.

> ## Highly-Available applications notes

Reviewer: ```suggestion
## Highly-Available Applications

Consider this in addition to the basic production environment requirements:
```

**Networking:**

- It is recommended to have multiple 10 Gbps (or faster) network cards.
  You can separate storage and application traffic by assigning them to different network interfaces.

> - Expect significant amount of horizontal, inter-node traffic inside clusters.
>   It is usually caused by multiple replicas of services and databases deployed across different nodes exchanging data.
>   Also, virtual machines with live migration require replicated volumes, which further increases the amount of traffic.

Reviewer: ```suggestion
- Expect a significant amount of horizontal, inter-node traffic inside clusters.
  It is usually caused by multiple replicas of services and databases deployed across different nodes exchanging data.
  Also, virtual machines with live migration require replicated volumes, which further increases the amount of traffic.
```
